mautic/mautic
# New EmailStatModel that will dispatch events on email Stat pre and post save

**Please be sure you are submitting this against the _staging_ branch.**

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | N
| New feature? | Y
| Automated tests included? | Y
| Related user documentation PR URL | /
| Related developer documentation PR URL | /
| Issues addressed (#s or URLs) | /
| BC breaks? | N
| Deprecations? | N

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:

We have a plugin that needs to make an action on email Stat entity save. The problem was that the Stat entity was saved through many places, sometimes through StatRepository, sometimes through EntityManager directly. 

This PR creates a new EmailStatModel that is dispatching new events on before and after the save and then refactors all the places where I could find was the Stat entity saved and used the new model instead.

This PR does not change any functionality. It should all still work as before.

Although there was 1 bug I discovered when I was covering a method I have changed with a test. The emails were not retried for 3 times before creating a DNC record but on the first retrial. It looks like someone forgot a debug `true` there.

All the changes are covered by unit tests. I cannot tell what different use cases needs to be tested.

#### Steps to test this PR:
1. Make sure you can send an email to a contact (or better multiple) and you can see in the contact detail page if you open the email or click a link.
2. Also test unsubscribtion.
3. And email reply events.

#### Other areas of Mautic that may be affected by the change:
1. All areas concerning sending email and recording changes on email stats like opens and retries.

#### List deprecations along with the new alternative:
1. None

#### List backwards compatibility breaks:
1. None
